contracts/test/StakeTable.t.sol

```diff
@@ -530,4 +553,162 @@ contract StakeTable_Test is Test {
+            assertGe(stakeTable.numPendingRegistrations(), 1);
+            return true;
+        }
+    }
+
+    ///@dev Test invariants about our queue logic holds during a random sequence of register,
+    /// requestExit, and advanceEpoch operations
+    function testFuzz_SequencesOfEvents(
+        uint8[ARRAY_SIZE] memory events,
+        uint8[ARRAY_SIZE] memory rands
+    ) external {
+        BN254.G2Point[ARRAY_SIZE] memory registeredKeys;
+
+        // Tracks the indices corresponding to an active key
+        bool[ARRAY_SIZE] memory isKeyActive;
+
+        uint64 numRegistrations = 0;
+        uint64 numExits = 0;
+
+        for (uint256 i = 0; i < ARRAY_SIZE; i++) {
+            uint256 ev = bound(events[i], 0, 2);
+
+            bool exitRequestSuccessful = false;
+
+            if (ev == 0) {
+                // Registrations
+                bool res = handleRegistrations(i, rands, registeredKeys, isKeyActive);
+                if (res) {
+                    numRegistrations++;
+                }
+            } else if (ev == 1) {
+                // Exits
+                if (numRegistrations == 0) {
+                    break;
+                }
+                uint256 indexRegistration = bound(rands[i], 0, numRegistrations - 1);
+                bytes32 hashNode = stakeTable._hashBlsKey(registeredKeys[indexRegistration]);
+                (
+                    address sender,
+                    AbstractStakeTable.StakeType stakeType,
+                    uint64 balance,
+                    uint64 registerEpoch,
+                    uint64 exitEpoch,
+                ) = stakeTable.nodes(hashNode);
+
+                balance;
+                stakeType;
+
+                BN254.G2Point memory blsVK = registeredKeys[indexRegistration];
+
+                bool canExit = (stakeTable.currentEpoch() >= registerEpoch + 1) && (exitEpoch == 0);
+                if (canExit) {
+                    vm.prank(sender);
+                    bool res = stakeTable.requestExit(blsVK);
+                    assertTrue(res);
+                    numExits++;
+                    exitRequestSuccessful = true;
+
+                    // Invariants specific to a successful exit
+                    assertGe(stakeTable.firstAvailableExitEpoch(), stakeTable.currentEpoch() + 1);
+                    assertGe(stakeTable.numPendingExits(), 1);
+                } else {
+                    vm.prank(sender);
+                    vm.expectRevert();
+                    bool res = stakeTable.requestExit(blsVK);
+                    assertFalse(res);
+                }
+            } else {
+                // Advance epoch
+                // ev == 2
+                uint64 currentEpoch = lightClientContract.currentEpoch();
+                uint64 nextEpoch = currentEpoch + 1;
+                lightClientContract.setCurrentEpoch(nextEpoch);
+                assertEq(stakeTable.currentEpoch(), nextEpoch);
+            }
+
+            // Global invariants
```

Comment:
I think another one would be, giving `numRegistration` and `Churn_rate`, we should calculate the first available epoch and pending queue size, we can check if contract also computed that correctly

Reply:
Finding such a general formula seems hard because epochs are not necessarily filled up with registrations as some epochs can be arbitrarily skipped.

Reply:
I see, what I describe may be a Registration-specific invariant, not a global one.
when I register, I first query the `_firstAvailable` and `pendingXX` variable, and do the calculation in the test, and then compare with what I actually got after I successfully register. 

- also, can we change visibility of `_firstAvailable` etc from `public` to `internal` instead?

Reply:
Here is  a way to make the "formula" work: The idea is to introduce a fuzzing variable `skipEpoch`. When `skipEpoch` is false, we don't skip the epochs and thus the queue of each epochs are filled up. This allows us to check the relationship between `numRegistration` , max churn rate and `pendingRegistrations` (same for next registration epoch). See 93264d8.

Reply:
> also, can we change visibility of _firstAvailable etc from public to internal instead?

We then would need a getter funcion?

Reply:
i think we can calculate w/o `skipEpoch` flag, 
simply by knowing the `currentEpoch`, `_firstAvailable` and `pendingXX` we should be able to compute the expected epoch for the new registration and the post-registration `pendingXX` value. 

there are two parts:
- logic of `nextRegistrationEpoch()`: given `currentEpoch`, `_firstAvail` and `_pendingXX`, does it return the expected value
- update of `_firstAvail` and `_pendingXX` after registration assuming a correct `nextRegistrationEpoch()`

To test the first, if we spell out the formula, it would be verbatim of the actual function, thus no need. Maybe we could add another test function with a few hardcoded instances (esp edge cases such as pending is just full, or the `firstAvail` is behind `currentEpoch`, and see if the output is expected

To test the second, we simply record down the output of `nextRegistrationEpoch()` before calling `register`, then inspect the variable again to see if they are properly updated.

^^ does this make sense?

Reply:
I am ok with implementing the second check. Now I think checking what happens when no skipping happens captures part of the behaviour of the queue.  Especially because it could happen in practice.

Reply:
> I am ok with implementing the second check

Done in ec56a87 for the registration queue. Is that what you had in mind?